From e712bedf14ee59e6ab6fb7f3ff9f9ea4bd12b3f1 Mon Sep 17 00:00:00 2001
From: Igor Grinberg <igrinberg@habana.ai>
Date: Thu, 05 Aug 2021 12:58:47 +0300
Subject: [PATCH] [SW-51853] gaudi: security: add dyn cfg block rrs

While in Preboot, Gaudi's engines are vulnerable to PCIe attacks and no
monitoring (e.g. interrupts) is done.
This can result in Gaudi to become unstable.
To overcome, add 2 PCIe RRs to dynamic configuration and block the
entire CFG space. Only the communication area is left open.

Change-Id: Idc6ad331890728589e8d9f1cb8973d0a01616e50
Signed-off-by: Igor Grinberg <igrinberg@habana.ai>
---

diff --git a/arch/gaudi/arm/security.c b/arch/gaudi/arm/security.c
index b88ad59..d46448e 100644
--- a/arch/gaudi/arm/security.c
+++ b/arch/gaudi/arm/security.c
@@ -1792,6 +1792,7 @@
 
 #define SRAM_SIZE	0x1400000	/* 20MB */
 #define SRAM_END	(SDRAM_BASE + SRAM_SIZE)
+#define HOST_PHYS_BASE	0x8000000000ull
 
 enum preboot_sec_rr_names {
 	SRAM_BGN_TO_VERS = 0,
@@ -1801,6 +1802,8 @@
 	IMG_END_TO_SRAM_END,
 	STM_SF_MEMMAP,
 	SPSRAM_IMAGE,
+	CFG_LOW,
+	CFG_HIGH,
 };
 
 struct preboot_sec_dyn_pcie_rr_cfg {
@@ -1844,6 +1847,16 @@
 		.rr = PCIE_DYN_RR_SPSRAM_IMG,
 		.min = FIT_IMG_SPSRAM_BASE - 1,
 		.max = FIT_IMG_SPSRAM_END,
+	},
+	[CFG_LOW] = {
+		.rr = PCIE_DYN_RR_CFG_LOW,
+		.min = HL_CFG_BASE - 1,
+		.max = mmPSOC_GLOBAL_CONF_BASE,
+	},
+	[CFG_HIGH] = {
+		.rr = PCIE_DYN_RR_CFG_HIGH,
+		.min = mmPSOC_GPIO0_BASE - 1,
+		.max = HOST_PHYS_BASE,
 	},
 };
 
@@ -1980,7 +1993,7 @@
 	preboot_secure_pcie_rr(PCIE_DYN_RR_SRAM_END, SDRAM_BASE - 1, SRAM_END);
 
 	/* Apply all already known configurations */
-	for (i = SRAM_BGN_TO_VERS; i <= SPSRAM_IMAGE; i++) {
+	for (i = 0; i < ARRAY_SIZE(preboot_sec_dyn_cfg); i++) {
 		cfg = &preboot_sec_dyn_cfg[i];
 
 		if (cfg->rr >= 0 && cfg->min && cfg->max)
